providers/ms_entra_id.go

```diff
@@ -0,0 +1,166 @@
+	multiTenantAllowedTenants []string
+
+	microsoftGraphURL *url.URL
+	isMultiTenant     bool
+}
+
+const (
+	microsoftEntraIDProviderName = "Microsoft Entra ID"
+)
+
+var (
+	microsoftGraphURL = &url.URL{
+		Scheme: "https",
+		Host:   "graph.microsoft.com",
+		Path:   "/v1.0/me",
+	}
+)
+
+// NewAzureProvider initiates a new MicrosoftEntraIDProvider
+func NewMicrosoftEntraIDProvider(p *ProviderData, opts options.Provider) *MicrosoftEntraIDProvider {
+	p.setProviderDefaults(providerDefaults{
+		name: microsoftEntraIDProviderName,
+	})
+
+	return &MicrosoftEntraIDProvider{
+		OIDCProvider: NewOIDCProvider(p, opts.OIDCConfig),
+
+		skipGraphGroups:           opts.MicrosoftEntraIDConfig.DisableGroupsFromGraph,
+		multiTenantAllowedTenants: opts.MicrosoftEntraIDConfig.MultiTenantAllowedTenants,
+		isMultiTenant:             strings.Contains(opts.OIDCConfig.IssuerURL, "common"),
+		microsoftGraphURL:         microsoftGraphURL,
+	}
+}
+
+// EnrichSession checks for group overage and passes through to generic EnrichSession()
+func (p *MicrosoftEntraIDProvider) EnrichSession(ctx context.Context, session *sessions.SessionState) error {
+	err := p.OIDCProvider.EnrichSession(ctx, session)
+
+	if !p.skipGraphGroups {
+		logger.Printf("checking for overage of azure groups")
+		hasGroupOverage, _ := p.checkGroupOverage(session)
+
+		if hasGroupOverage {
+			logger.Printf("overage found, reading groups from graph")
+			if err = p.addGraphGroupsToSesion(ctx, session); err != nil {
+				return fmt.Errorf("unable to read groups from graph: %v", err)
+			}
+		}
+	}
+
+	return err
+}
+
+// ValidateSession checks for allowed tenants for multi-tenant apps and passes through to generic ValidateSession
+func (p *MicrosoftEntraIDProvider) ValidateSession(ctx context.Context, session *sessions.SessionState) bool {
+
+	if p.isMultiTenant {
+		issuer, exists, error := p.getIssuer(session)
+		if issuer == "" || !exists || error != nil {
+			return false
+		}
+
+		if p.multiTenantAllowedTenants != nil {
+			logger.Printf("List of allowed tenant is specified, we check if %s is allowed issuer", issuer)
+			tenantAllowed := p.checkIssuerMatchesTenantList(issuer, p.multiTenantAllowedTenants)
+			if !tenantAllowed {
+				return false
+			}
+		}
+	}
+
+	return p.OIDCProvider.ValidateSession(ctx, session)
+}
+
+// checkGroupOverage checks is claims present in ID Token indicate group overage
+func (p *MicrosoftEntraIDProvider) checkGroupOverage(session *sessions.SessionState) (bool, error) {
+	extractor, err := p.getClaimExtractor(session.IDToken, session.AccessToken)
+	if err != nil {
+		return false, fmt.Errorf("unable to get claim extractor: %v", err)
+	}
+
+	claimNames, _, _ := extractor.GetClaim("_claim_names")
+	hasGroups, _, _ := extractor.GetClaim("hasgroups")
+
+	if claimNames == nil && hasGroups == nil {
+		return false, nil
+	}
+
+	var claimNamesMap map[string]string
+
+	if claimNames != nil {
+		claimNamesMap = cast.ToStringMapString(claimNames)
```

Comment:
is it necessary to use an external dependency for this remapping?

Reply:
What we check here is a following claim:
```
"_claim_names": {
  "groups": "src1"
}
```
So I need to somehow transform content of a claim it to `map[string]string`. I must have found this library in [claim_extractor.go](https://github.com/oauth2-proxy/oauth2-proxy/blob/master/pkg/providers/util/claim_extractor.go#L15). Any better method appreciated :smile: